src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.PackageManagement.VisualStudio;
+using NuGet.Packaging;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.VisualStudio.Contracts;
+
+namespace NuGet.VisualStudio.Implementation.Extensibility
+{
+    internal class NuGetProjectService : INuGetProjectService
+    {
+        private readonly IVsSolutionManager _solutionManager;
+
+        public NuGetProjectService(IVsSolutionManager solutionManager)
+        {
+            _solutionManager = solutionManager ?? throw new ArgumentNullException(nameof(solutionManager));
+        }
+
+        public async Task<InstalledPackagesResult> GetInstalledPackagesAsync(Guid projectId, CancellationToken cancellationToken)
```

Comment:
The one tiny concern I have is code duplication from the current project services implementation. 
This is a lot cleaner, but it does give me a bit of a pause.

Reply:
You're referring to [`VsPackageInstallerServices.GetInstalledPackages(DTE.Project)`](https://github.com/NuGet/NuGet.Client/blob/070c0166678caf816d4bcf7c9a4d2b168287e1d0/src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstallerServices.cs#L161)

There are some similarities:
* using `NuGetProject.GetInstalledPackages()` to get the packages list installed in the project
* iterating the list of `PackageReference` objects to convert to the method's return type
* Trying to find the package's install path.

There are some differences:
* Method input (`DTE.Project` vs `System.Guid`), and therefore how to get the `NuGetProject` instance
* Method return type (`IVsPackageMetadata` vs `NuGetInstalledPackage`)
* This implementation wants to know if the project has nomination errors, to let the caller know that the project is in a bad state, even if some packages are returned

However, looking carefully at the `VsPackageInstallerServices.GetInstalledPackages(DTE.Project)` and `VsPackageInstallerServices.GetInstalledPackages()`, I see my implementation is inconsistent with them in  two ways:
* `VsPackageInstallerServices` skips packages with floating versions, so the caller doesn't even know the package is installed.
  * I don't feel like we should reproduce this behavior in the new service. Improving `VsPackageInstallerServices` to support floating versions feels very much out of scope for my work.
* `VsPackageInstallerServices` assumes that any project that does not extend `BuildIntegratedNuGetProject` is uses the solution packages folder. In order to minimise the number of MEF services needed to get the `FolderNuGetProject` (solution packages folder), my implementation explicitly checks for `BuildIntegratedNuGetProject` (package reference) and `MSBuildNuGetProject` (packages.config), with a fallthrough if the project is neither.
  * Looking at classes that implement `NuGetProject`, only `ProjectKNuGetProjectBase`, `PackagesConfigNuGetProject` and `FolderNuGetProject` don't extend either `BuildIntegratedNuGetProject` or `MSBuildNuGetProject`. Two of them don't make sense in the context of a Visual Studio project, and Project K projects I don't believe have been supported in VS for years now (VS used to automatically migrate it to a csproj. If that doesn't happen any longer, then Project K projects aren't supported at all).

I understand the sentiment of your comment. `VsPackageInstallerServices.GetInstalledPackages` and `NuGetProjectService.GetInstalledPackages` conceptually return the same information, just in different formats and using different call patterns, so ideally should reuse much of the same implementation. My response is that the core logic is already re-used via `NuGetProject.GetInstalledPackages`, `FolderNuGetProject.GetInstalledPath` and `FallbackPackagePathResolver .GetPackageDirectory`. The rest doesn't feel that important to me.

Reply:
Makes sense.

The current implementation is *a lot* cleaner.

Reply:
A good set of tests added in a future PR would offset many of my concerns :)